### DISCUSSION TITLE
Smart contract supporting MiMC hash in gnark

### DISCUSSION THREAD

**Discussion by heianqishijueqile** [Consensys/gnark#1252] [General]
In Tornado Cash, there is a contract specifically for computing the MiMC hash, which can be used with Circom to create zero-knowledge proofs. If I want to use gnark for MiMC hash zero-knowledge proofs, is there a contract compatible with the MiMC hash algorithm in gnark?

**Comment by ThomasPiellard** [comment]
hi @heianqishijueqile , we do have the implementation of mimc on bls12-377 (corresponding to gnark-crypto's implementation) here https://github.com/Consensys/linea-monorepo/blob/main/contracts/contracts/lib/Mimc.sol

We unrolled the loop for performances reason, you can adapt the code for the implementation of mimc on other curves (it's the exact same loop, only the counter and the constants differ)

**Reply by cxmqq333** [reply]
> hi @heianqishijueqile , we do have the implementation of mimc on bls12-377 (corresponding to gnark-crypto's implementation) here https://github.com/Consensys/linea-monorepo/blob/main/contracts/contracts/lib/Mimc.sol
> 
> We unrolled the loop for performances reason, you can adapt the code for the implementation of mimc on other curves (it's the exact same loop, only the counter and the constants differ)

"Is there an implementation based on other curves?" like bn254 bls12-381 ....

**Reply by ivokub** [reply]
We don't have a direct implementation. But modifying the Solidity code from bls12-377 MiMC to other curves should be straightforward by referencing the round constants using [`GetConstants()`](https://pkg.go.dev/github.com/consensys/gnark-crypto@v0.14.0/ecc/bn254/fr/mimc#GetConstants) method.

**Reply by cxmqq333** [reply]
We have just implemented the mimic based on bls12-381. Do we need to submit a PR to merge it into the demo?

**Reply by ivokub** [reply]
I think it is not necessary - you could use your local fork. Linea codebase is not general purpose library and the PR will probably be rejected.

**Comment by Makubu** [comment]
I have tried to implement MiMC on BN254  using the Consensys implementation on [BLS12-377.](https://github.com/Consensys/linea-monorepo/blob/439cc10800c44cd6f91ddc05a2965107db35ad5a/contracts/src/libraries/Mimc.sol) I changed the header (from bytes to uint256). The version of BLS12-377 works with my test workflow but no the BN254 version. I have just taken the same code, add rounds to fit all the constants of BN254 and changed the field size. Is there another modification to make that i have not done ?
